src/dftbp/dftb/mdftb.F90

```diff
@@ -0,0 +1,1575 @@
+    procedure :: addMultiExpanGradients
+    procedure :: addAtomicDipoleMoment
+    procedure :: addAtomicQuadrupoleMoment
+    procedure :: getMultiExpanInfo
+    procedure :: getOrbitalEquiv
+  end type TMdftb
+
+  !> Number of dipole components used in tblite library (x, y, z)
+  integer, parameter :: dimDipole = 3
+
+  !> Number of quadrupole components used in tblite library (xx, xy, yy, xz, yz, zz)
+  integer, parameter :: dimQuadrupole = 6
+
+contains
+
+  !> Get information on required multipolar contributions
+  subroutine getMultiExpanInfo(this, nDipole, nQuadrupole)
+
+    !> Data structure
+    class(TMdftb), intent(in) :: this
+
+    !> Number of dipole moment components
+    integer, intent(out) :: nDipole
+
+    !> Number of quadrupole moment components
+    integer, intent(out) :: nQuadrupole
+
+    integer, parameter :: dimDipole = 3
+    integer, parameter :: dimQuadrupole = 6
+
+    nDipole = dimDipole
+    nQuadrupole = dimQuadrupole
+
+  end subroutine getMultiExpanInfo
+
+
+  !> Returns the equivalence to get the correct mixing of charge dependent contributions
+  subroutine getOrbitalEquiv(this, equivDip, equivQuad)
+
+    !> Data structure
+    class(TMdftb), intent(inout) :: this
+
+    !> The equivalence vector for cumulative atomic dipole populations
+    integer, intent(out) :: equivDip(:,:)
+
+    !> The equivalence vector for cumulative atomic quadrupole populations
+    integer, intent(out) :: equivQuad(:,:)
+
+    integer :: nAtom, iCount, iSpin, nSpin, iAt, iSp, ii
+
+    nAtom = size(equivDip, dim=2)
+    nSpin = 1
+
+    equivDip(:,:) = 0
+    equivQuad(:,:) = 0
+
+    ! equivDip
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimDipole
+        iCount = iCount + 1
+        equivDip(ii, iAt) = iCount
+      end do
+    end do
+
+    ! equivQuad
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimQuadrupole
+        iCount = iCount + 1
+        equivQuad(ii, iAt) = iCount
+      end do
+    end do
+
+  end subroutine getOrbitalEquiv
+
+
+  !> Initializes instance.
+  subroutine TMdftb_init(this, inp)
+
+    !> Instance.
+    type(TMdftb), intent(out) :: this
+
+    !> Input data.
+    type(TMdftbInp), intent(in) :: inp
+
+    integer, parameter :: icx = 1, icy = 2, icz = 3
+    integer, parameter :: ios = 0
+    integer, parameter :: iopy = 0, iopz = 1, iopx = 2
+    integer, parameter :: iodxy = 0, iodyz = 1, iodzz = 2, iodxz = 3, iodxxyy = 4
+    real(dp), parameter :: minIntgrl = 1.0e-9_dp
+    integer :: nAtoms, mOrb, nSpecies, iAt1, iAt2, nOrb1, nOrb2, ii, jj, iSp1, iSp2
+    integer :: mu, nu, mm, nn
+    integer :: ang1, ang2, iSh1, iSh2, iOrbAng1, iOrbAng2
+    real(dp) tmpIntgrl, tmpAvgTrace
+
+    this%nAtoms = size(inp%orb%nOrbAtom)
+    this%nSpecies = inp%nSpecies
+    this%nOrb = inp%nOrb
+    this%mOrb = inp%orb%mOrb
+    this%nSpin = inp%nSpin
+    this%hubbu = inp%hubbu(:)
+    this%species = inp%species(:)
+    this%nOrbSpecies = inp%orb%nOrbSpecies(:)
```

Comment:
```suggestion
    this%hubbu = inp%hubbu
    this%species = inp%species
    this%nOrbSpecies = inp%orb%nOrbSpecies
```
We do not use dummy slice operators on the *right* hand side. (But always on the *left* hand side, if no reallocation is required.)